Add support for multiple ratings

phate89 opened the pull request:
Since https://github.com/xbmc/xbmc/pull/8062 is strictly related to this change I kept working on this so I can have an opinion about my solution.
I moved rating and votes to a separate table (so we can have more than one for every movie) and I added the rating_id of the default vote to the movie table (using the old rating column). In CVideoInfoTag class all the ratings are stored in a map and they're handled by functions..

Montellese commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -4396,23 +4400,25 @@ std::string CGUIInfoManager::GetVideoLabel(int item)
       break;
     case VIDEOPLAYER_RATING_AND_VOTES:
-      {
+    {
```
"Bad" indentation.

Montellese commented on xbmc/interfaces/json-rpc/VideoLibrary.cpp:
```diff
@@ -1123,8 +1123,25 @@ void CVideoLibrary::UpdateVideoTag(const CVariant &parameterObject, CVideoInfoTa
 
   if (ParameterNotNull(parameterObject, "track"))
     details.m_iTrack = (int)parameterObject["track"].asInteger();
-  if (ParameterNotNull(parameterObject, "rating"))
```
Can we keep `rating` and `votes` support here which will simply modify the default rating/votes? That way the API would stay backwards compatible.

Montellese commented on xbmc/interfaces/json-rpc/VideoLibrary.cpp:
```diff
@@ -1123,8 +1123,25 @@ void CVideoLibrary::UpdateVideoTag(const CVariant &parameterObject, CVideoInfoTa
+  if (ParameterNotNull(parameterObject, "ratings"))
+  {
+    CVariant r = parameterObject["ratings"];
+    for (CVariant::const_iterator_map rIt = r.begin_map(); rIt != r.end_map(); rIt++)
```
`for (const auto& rIt = r.begin_map(); ...)`

Montellese commented on xbmc/interfaces/json-rpc/VideoLibrary.cpp:
```diff
@@ -1123,8 +1123,25 @@ void CVideoLibrary::UpdateVideoTag(const CVariant &parameterObject, CVideoInfoTa
-    details.m_fRating = parameterObject["rating"].asFloat();
+  if (ParameterNotNull(parameterObject, "ratings"))
+  {
+    CVariant r = parameterObject["ratings"];
```
`r` is not a very meaningful variable name.

Montellese commented on xbmc/interfaces/json-rpc/VideoLibrary.cpp:
```diff
@@ -1123,8 +1123,25 @@ void CVideoLibrary::UpdateVideoTag(const CVariant &parameterObject, CVideoInfoTa
+      details.SetRating(rIt->second["rating"].asFloat(), rIt->second["name"].asString());
+      if (ParameterNotNull(rIt->second, "votes"))
+        details.SetVotes(StringUtils::ReturnDigits(parameterObject["votes"].asString()), rIt->second["name"].asString());
+    }
```
Bad indentation of the whole block.

Montellese commented on xbmc/utils/GroupUtils.cpp:
```diff
@@ -100,16 +100,17 @@ bool GroupUtils::Group(GroupBy groupBy, const std::string &baseDir, const CFileI
       setInfo->m_strTitle = pItem->GetLabel();
 
       int ratings = 0;
+      float tot = 0;
```
Please don't abbreviate your variable names to much. It makes the code less obvious.

Montellese commented on xbmc/utils/GroupUtils.cpp:
```diff
@@ -139,8 +140,7 @@ bool GroupUtils::Group(GroupBy groupBy, const std::string &baseDir, const CFileI
 
       if (ratings > 1)
-        pItem->GetVideoInfoTag()->m_fRating /= ratings;
-
```
Unnecessary whitespace change.

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -199,6 +200,9 @@ void CVideoDatabase::CreateTables()
   m_pDS->exec("CREATE TABLE tag_link (tag_id integer, media_id integer, media_type TEXT)");
+
+  CLog::Log(LOGINFO, "create rating table");
+  m_pDS->exec("CREATE TABLE rating(rating_id INTEGER PRIMARY KEY, media_id INTEGER, media_type TEXT, rating_type TEXT, rating FLOAT, votes INTEGER)");
```
Space between `rating` and `(rating_id ...)`

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -444,25 +460,33 @@ void CVideoDatabase::CreateViews()
+                                      "    rating.rating_id=movie.c%02d",
+                                      VIDEODB_ID_RATING_ID);
+  m_pDS->exec(movieview.c_str());
 }
```
Are there no ratings for musicvideos?

phate89 replied:
I tried to look but seems they haven't

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -1385,6 +1409,48 @@ int CVideoDatabase::AddToTable(const std::string& table, const std::string& firs
+  {
+    if (NULL == m_pDB.get()) return -1;
+    if (NULL == m_pDS.get()) return -1;
+    if (values.size() > 0)
```
Please always use curly braces around multi-line blocks. But you don't really need the `values.size() > 0` check (which should be `!values.empty()`) because the loop won't do anything on an empty map.

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -1385,6 +1409,48 @@ int CVideoDatabase::AddToTable(const std::string& table, const std::string& firs
+    if (NULL == m_pDB.get()) return -1;
+    if (NULL == m_pDS.get()) return -1;
+    if (values.size() > 0)
+    for (RatingMap::const_iterator i = values.begin(); i != values.end(); i++)
```
`for (const auto& rating : values)`

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -1386,6 +1407,48 @@ int CVideoDatabase::AddToTable(const std::string& table, const std::string& firs
   return -1;
 }
 
+int CVideoDatabase::AddRatings(int mediaId, const char *mediaType, const RatingMap& values, const std::string& default)
```
`default` is a reserved keyword: http://en.cppreference.com/w/cpp/keyword/default

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -1386,6 +1407,48 @@ int CVideoDatabase::AddToTable(const std::string& table, const std::string& firs
+        strSQL = PrepareSQL("UPDATE rating SET rating.rating = '%.f', rating.votes = %i WHERE rating.media_id = %i", i->second.rating, i->second.votes, id);
+        m_pDS->exec(strSQL.c_str());        
+      }
+      if (i->first == default)
```
Is this really needed if the `CRating` structure has a default flag?

phate89 replied:
CRating hasn't a default flag. You mean I can add it to avoid this?

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -2432,6 +2503,9 @@ int CVideoDatabase::SetDetailsForEpisode(const std::string& strFilenameAndPath,
     AddActorLinksToItem(idEpisode, MediaTypeEpisode, "writer", details.m_writingCredits);
 
+    // add ratings
+    details.m_iIndexRating = AddRatings(idEpisode, MediaTypeEpisode, details.m_rating, details.m_strDefaultRating);
```
Hm I'm not sure if we should do this here (same for tvshows etc). We also don't do it for movie sets.

phate89 replied:
We don't do it for movie sets because idset is not a cXX field so is manually added to the query directly with the new id. cXX fields are loaded by a function that loops automatically and I can't change that part of the query. The alternative is to add another "rating_id" column and leave another cXX variable unused

Montellese commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -3778,6 +3864,28 @@ void CVideoDatabase::GetTags(int media_id, const std::string &media_type, std::v
+    if (!m_pDB.get()) return;
+    if (!m_pDS2.get()) return;
+
+    std::string sql = PrepareSQL("SELECT rating.rating_type, rating.rating, rating.votes FROM rating WHERE rating.media_id = %i AND rating.media_type = '%s' ORDER BY rating.rating_id", media_id, media_type.c_str());
```
Do we really need the `ORDER BY` clause?

Montellese commented on xbmc/video/VideoDatabase.h:
```diff
@@ -83,6 +83,9 @@ namespace VIDEO
 #define VIDEODB_DETAILS_MOVIE_TOTAL_TIME        VIDEODB_MAX_COLUMNS + 12
+#define VIDEODB_DETAILS_MOVIE_RATING            VIDEODB_MAX_COLUMNS + 13
+#define VIDEODB_DETAILS_MOVIE_VOTES             VIDEODB_MAX_COLUMNS + 14
+#define VIDEODB_DETAILS_MOVIE_DEFAULT_RATING    VIDEODB_MAX_COLUMNS + 15
```
The column name is `rating_type` which doesn't match `DEFAULT_RATING` unless I misunderstood something.

phate89 replied:
I changed because in the view it changes the meaning. It's not the rating type anymore but it's the default rating type preferred for the media. I'll change it to VIDEODB_DETAILS_MOVIE_DEFAULT_RATING_TYPE

Montellese replied:
IMO the name just needs to state that it's a type and not a rating.

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -113,7 +113,22 @@ bool CVideoInfoTag::Save(TiXmlNode *node, const std::string &tag, bool savePathI
   if (!m_strSortTitle.empty())
     XMLUtils::SetString(movie, "sorttitle", m_strSortTitle);
-  XMLUtils::SetFloat(movie, "rating", m_fRating);
+  if (m_rating.size()>0)
```
`if (!m_rating.empty())`

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -113,7 +113,22 @@ bool CVideoInfoTag::Save(TiXmlNode *node, const std::string &tag, bool savePathI
+  {
+    TiXmlElement ratings("ratings");
+    RatingMap::iterator it;
+    for (it = m_rating.begin(); it != m_rating.end(); it++)
```
`for (const auto& it : m_rating)`

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -320,7 +333,12 @@ void CVideoInfoTag::Archive(CArchive& ar)
     ar << m_strUniqueId;
-    ar << m_fRating;
+    RatingMap::iterator i;
+    for (i = m_rating.begin(); i != m_rating.end(); i++)
```
`for (const auto& it : m_rating)`

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -320,7 +333,12 @@ void CVideoInfoTag::Archive(CArchive& ar)
     ar << m_iEpisode;
     ar << m_strUniqueId;
-    ar << m_fRating;
+    RatingMap::iterator i;
```
You first need to store the number of rating items that exist and then serialize the ratings (with all their details).

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -434,7 +456,7 @@ void CVideoInfoTag::Serialize(CVariant& value) const
   value["plot"] = m_strPlot;
   value["title"] = m_strTitle;
-  value["votes"] = m_strVotes;
+  value["votes"] = StringUtils::Format("%i", GetRating().votes);
```
std::to_string?

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -475,7 +497,19 @@ void CVideoInfoTag::Serialize(CVariant& value) const
+  value["rating"] = GetRating().rating;
+  value["ratings"] = CVariant(CVariant::VariantTypeArray);
+  RatingMap::const_iterator i;
+  for (i = m_rating.begin(); i != m_rating.end(); i++)
```
`for (const auto& it : m_rating)`

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -569,6 +603,14 @@ void CVideoInfoTag::ToSortable(SortItem& sortable, Field field) const
+{
+  if (type.empty())
+    return m_rating.find(m_strDefaultRating)->second;
+  else
```
No need for the `else`

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -600,15 +642,42 @@ void CVideoInfoTag::ParseNative(const TiXmlElement* movie, bool prioritise)
   XMLUtils::GetInt(movie, "userrating", m_iUserRating);
   XMLUtils::GetFloat(movie, "epbookmark", m_fEpBookmark);
   int max_value = 10;
-  const TiXmlElement* rElement = movie->FirstChildElement("rating");
```
Could we keep support for `rating` and `votes` in case there's no `ratings` tag to still support existing NFOs?

phate89 replied:
It already works like this. if i don't find ratings i search for rating and I set it as "default" rating

Montellese commented on xbmc/video/VideoInfoTag.h:
```diff
@@ -47,6 +47,16 @@ struct SActorInfo
+class CRating
+{
+public:
+  CRating() { CRating(0.0, 0); }
```
I'm pretty sure that this doesn't do what you expect. Please use constructor initialization like
```
CRating()
  : CRating(0.0f, 0)
{ }
```

Montellese commented on xbmc/video/VideoInfoTag.h:
```diff
@@ -106,7 +117,10 @@ class CVideoInfoTag : public IArchivable, public ISerializable, public ISortable
   void SetSortTitle(std::string sortTitle);
   void SetPictureURL(CScraperUrl &pictureURL);
-  void SetVotes(std::string votes);
+  void AddRating(float rating, int votes, std::string type = "");
```
`const std::string& type = "")` please (same below).

Montellese commented on xbmc/video/VideoInfoTag.h:
```diff
@@ -175,7 +189,9 @@ class CVideoInfoTag : public IArchivable, public ISerializable, public ISortable
   int m_iSpecialSortEpisode;
   int m_iTrack;
-  float m_fRating;
+  RatingMap m_rating;
```
Please name it `m_ratings` to make it clear that there's more than one.

Montellese commented on xbmc/video/VideoInfoTag.h:
```diff
@@ -175,7 +189,9 @@ class CVideoInfoTag : public IArchivable, public ISerializable, public ISortable
   int m_iTrack;
-  float m_fRating;
+  RatingMap m_rating;
+  int m_iIndexRating;
```
We use `Id` and not `Index` so please rename to `m_iIdDefaultRating`.

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -998,9 +1065,32 @@ void CVideoInfoTag::SetPictureURL(CScraperUrl &pictureURL)
 }
 
-void CVideoInfoTag::SetVotes(std::string votes)
+void CVideoInfoTag::AddRating(float rating, int votes, std::string type)
```
Please specify the default value (same below): `const std::string& type /* = "" */`

Montellese commented:
A few things that come to mind:
* you need to split the changes to libUPnP into it's own commit again and add a commit with the patch
* there doesn't seem to be a way to delete ratings
* this needs a JSON-RPC API version bump

phate89 commented:
thanks @Montellese . I fixed them and I also fixed the json-rpc error and bumped the version.
About deleting them I don't see a better solution than delete every rating for the item every update. I'll add a trigger to delete them once the media is deleted soon

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -600,15 +652,42 @@ void CVideoInfoTag::ParseNative(const TiXmlElement* movie, bool prioritise)
+  {
+    CRating r;
+    XMLUtils::GetFloat(movie, "rating", r.rating);
+	if (XMLUtils::GetString(movie, "votes", value))
```
Odd indentation.

phate89 committed:
add support for multiple ratings

phate89 commented:
updated again. added triggers for rating removal and created libUPNP patch in a separate commit

phate89 committed:
platinum: add support for multiple ratings

phate89 commented:
I removed WIP since my problems i stated were all solved. If you find some other missing part please  report it

phate89 commented:
rebased..

MartijnKaijser commented:
jenkins build this please

Montellese commented:
There are already conflicts again due to the userrating PR having been merged :-/

Montellese commented:
Haven't looked at the whole diff anymore as I've reviewed it before. Just looked at the videodb and SQL stuff. Found a few minors but apart from that (and needing a rebase) it should be good. The additional ratings aren't shown anywhere in the GUI yet right?

phate89 commented:
@Montellese sorry but can you point me the lines your comments are referring to? 
yes the ratings right now aren't shown to the gui but scrapers can add them to the db and skinners can use them..

Montellese commented:
@phate89: Just click on the file name in the header of my comments and it will take you to the proper file and line with the comment.

phate89 commented:
thanks I missed that. I removed the checks on clean database, added try/catch stuff and rebased (again)..

Montellese commented:
jenkins build this please

Razzeee commented on xbmc/video/VideoDatabase.h:
```diff
@@ -827,6 +836,8 @@ class CVideoDatabase : public CDatabase
 
   int AddToTable(const std::string& table, const std::string& firstField, const std::string& secondField, const std::string& value);
+  int UpdateRatings(int mediaId, const char *mediaType, const RatingMap& values, const std::string& defaultRating);
+  int AddRatings(int mediaId, const char *mediaType, const RatingMap& value, const std::string& default);
```
Other platforms then win32 seem to not like a variable named 'default'

phate89 commented:
@Razzeee that's because 'default' is a new feature in c++ 11. I wonder why it worked under windows.. btw it's fixed now..

Razzeee commented:
Thanks.
So will the old scrapers still work or will they break until they are aligned with this?

jenkins build this please

phate89 commented:
They works. Of course since I don't know the rating source it's called 'default'

Montellese commented on xbmc/interfaces/json-rpc/VideoLibrary.cpp:
```diff
@@ -1124,7 +1124,24 @@ void CVideoLibrary::UpdateVideoTag(const CVariant &parameterObject, CVideoInfoTa
+  {
+    CVariant ratings = parameterObject["ratings"];
+    for (auto& rIt = ratings.begin_map(); rIt != ratings.end_map(); rIt++)
+      if (rIt->second.isArray() && ParameterNotNull(rIt->second, "name") && ParameterNotNull(rIt->second, "rating"))
```
Please use curly braces on any multi-line scopes.

Montellese commented on xbmc/utils/StringUtils.cpp:
```diff
@@ -512,6 +512,15 @@ std::string& StringUtils::TrimRight(std::string &str, const char* const chars)
+int StringUtils::ReturnDigits(std::string str)
+{
+  std::stringstream ss;
+  for (unsigned int i = 0; i < str.size(); i++)
```
Curly braces.

Montellese replied:
Add use `for (const auto& character : str)` to loop through the string.

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -114,7 +114,22 @@ bool CVideoInfoTag::Save(TiXmlNode *node, const std::string &tag, bool savePathI
   if (!m_strSortTitle.empty())
     XMLUtils::SetString(movie, "sorttitle", m_strSortTitle);
-  XMLUtils::SetFloat(movie, "rating", m_fRating);
+  if (m_ratings.empty())
```
You're missing a `!` i.e. `if (!m_ratings.empty())`.

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -114,7 +114,22 @@ bool CVideoInfoTag::Save(TiXmlNode *node, const std::string &tag, bool savePathI
+  if (m_ratings.empty())
+  {
+    TiXmlElement ratings("ratings");
+    RatingMap::iterator it;
```
Not needed.

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -329,7 +342,15 @@ void CVideoInfoTag::Archive(CArchive& ar)
     ar << m_strUniqueId;
-    ar << m_fRating;
+    ar << (int)m_ratings.size();
+    RatingMap::iterator i;
```
Not needed.

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -414,7 +434,21 @@ void CVideoInfoTag::Archive(CArchive& ar)
-    ar >> m_fRating;
+    int iRatingSize;
+    ar >> iRatingSize;
+    for (int i = 0; i<iRatingSize; ++i)
```
Spaces around `<`.

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -604,6 +649,13 @@ void CVideoInfoTag::ToSortable(SortItem& sortable, Field field) const
+const CRating CVideoInfoTag::GetRating(std::string type) const
+{
+  if (type.empty())
+    return m_ratings.find(m_strDefaultRating)->second;
```
I'd rather call `GetRating()` again with `m_strDefaultRating` to avoid code duplication.

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -604,6 +649,13 @@ void CVideoInfoTag::ToSortable(SortItem& sortable, Field field) const
+{
+  if (type.empty())
+    return m_ratings.find(m_strDefaultRating)->second;
+  return m_ratings.find(type)->second;
```
You need to handle the case where the call to `find` fails or it will go sideways when someone tries to get a rating for a non-existing type.

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -635,15 +687,42 @@ void CVideoInfoTag::ParseNative(const TiXmlElement* movie, bool prioritise)
+  const TiXmlElement* node = movie->FirstChildElement("ratings");
+  if (node)
+  {
+    for (const TiXmlElement* child = node->FirstChildElement("rating"); child != NULL; child = child->NextSiblingElement("rating"))
```
`nullptr` instead of `NULL`.

Montellese commented on xbmc/video/VideoInfoTag.cpp:
```diff
@@ -635,15 +687,42 @@ void CVideoInfoTag::ParseNative(const TiXmlElement* movie, bool prioritise)
+    CRating r;
+    XMLUtils::GetFloat(movie, "rating", r.rating);
+	  if (XMLUtils::GetString(movie, "votes", value))
+		  r.votes = StringUtils::ReturnDigits(value);
```
Bad indentation.

Montellese commented:
Took another look and found some more minors and some bugs.

Razzeee commented on xbmc/interfaces/json-rpc/VideoLibrary.cpp:
```diff
@@ -1124,7 +1124,24 @@ void CVideoLibrary::UpdateVideoTag(const CVariant &parameterObject, CVideoInfoTa
+  if (ParameterNotNull(parameterObject, "ratings"))
+  {
+    CVariant ratings = parameterObject["ratings"];
+    for (auto& rIt = ratings.begin_map(); rIt != ratings.end_map(); rIt++)
```
The other platform compilers don't like this :(
http://jenkins.kodi.tv/job/LINUX-RBPI/5305/console

Montellese replied:
`const auto& rIt`.

phate89 replied:
with const auto& rIt it will fail the ++ operand...

phate89 commented:
fixed everything.. for xbmc\interfaces\json-rpc\VideoLibrary.cpp on line 1136 I went back to CVariant::const_iterator_map

Razzeee commented:
jenkins build this please

phate89 commented:
Someone saw why they failed to build in android only?

MartijnKaijser commented:
Let's try again so you can look jenkins build this please

phate89 commented:
ok I used c++11 `std::to_string` but it seems that android likes it only if you force c++11. Now it should be ok

MartijnKaijser commented:
jenkins build this please

Razzeee commented:
Can you rebase this?

phate89 commented:
rebased

MartijnKaijser commented on xbmc/interfaces/json-rpc/schema/types.json:
```diff
@@ -510,7 +510,7 @@
     "items": { "type": "string",
       "description": "Requesting the cast, showlink and/or tag field will result in increased response times",
-      "enum": [ "title", "genre", "year", "rating", "director", "trailer",
+      "enum": [ "title", "genre", "year", "rating", "ratings", "director", "trailer",
```
@Montellese ping. Don' t we require to add new stuff at the end?

Montellese replied:
Yes please.

MartijnKaijser commented:
Missing the json-rpc version bump and our video db is already at 100 now so that bump is also missing.
Are there any changes to python API as well (or missing)?
http://mirrors.kodi.tv/docs/python-docs/16.x-jarvis/xbmcgui.html#ListItem-setInfo

phate89 commented:
I already fixed the db change. About the python api I didn't change them. RIght now they set the value to the default one. I'll bump api version

MartijnKaijser commented on xbmc/interfaces/json-rpc/schema/version.txt:
```diff
@@ -1 +1 @@
-6.33.0
+6.33.1
```
need to be 6.34.0 as this is a feature

phate89 replied:
fixed.thanks

MartijnKaijser commented:
changing ListItem.cpp also touches https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/legacy/ListItem.h#L174-L251
Maybe @tamland has more insight in this part

tamland commented:
Uhm, there are no changes to python api. It would be _nice_ to have python support for it as well, but that's not a requirement for adding a feature.

tamland commented on xbmc/utils/StringUtils.cpp:
```diff
@@ -512,6 +512,17 @@ std::string& StringUtils::TrimRight(std::string &str, const char* const chars)
+  }
+  return atoi(ss.str().c_str());
+}
+
```
Not sure what you're trying to do here. It extracts all integers from the string, then return 0 on error anyway. Simply `atoi` should be sufficient, right?

phate89 replied:
Votes field was string and that string could be formatted in several ways, for example
10000
10.000
10'000
10.000 votes
That's the safest way to get the votes. Here's the discussion:
https://github.com/xbmc/xbmc/pull/8062

tamland replied:
Ok, sorry for the noise.

phate89 commented:
@tamland 
can you point me to an implementation of a similar function? I'm not an expert of kodi code but I'll add python support if I can

tamland commented on xbmc/utils/StringUtils.cpp:
```diff
@@ -512,6 +512,17 @@ std::string& StringUtils::TrimRight(std::string &str, const char* const chars)
   return str;
 }
 
+int StringUtils::ReturnDigits(std::string str)
```
This should be const std::string& though..

tamland commented:
@phate89 Like f464cf7919b370c4b6 for instance? It's just documentation + converting the key/value to a VideoInfoTag property.

phate89 commented:
@tamland  thanks.. There is no way with current code to have a tuple with 3 elements right? Because to fill the rating you need 3 elements (vote, nr of votes, rating name)

phate89 commented:
Moved ratings in json types to the end

tamland commented:
@phate89 We have a 3-tuple type, but I don't think its possible to extract a tuple from setInfo like that (or it would bloat the InfoLabelDict type even more by adding another Alternative). Probably easier and better to add a `setRating` method if you need complex types like that.

Razzeee commented:
How to proceed?

tamland commented:
As I said all python parts can be done later. Imo just rebase and merge, and don't let that block this pr any longer.

phate89 committed:
platinum: add patch for 77ba91ac8815c150c7e4537a3622df52bd811cef

phate89 commented:
rebased. I agree.. Plus there's the editing via json-rpc if an addon wants to edit ratings

Razzeee commented:
jenkins build and merge

ronie commented:
am i correct in understanding this adds the following for skinners:
- ListItem.Property(Rating.n)
- ListItem.Property(Votes.n)

but not:
- ListItem.Property(RatingAndVotes.n)

?

phate89 commented:
yes I didn't think about that possibility. Is it useful right? If it is I'll add it

phate89 commented:
Wait if I'm not wrong is:
Listitem.Rating(n)

ronie commented:
sure, it would be nice to have. many skins use RatingAndVotes.

one more question, how do we know what those ratings represent?
wouldn't it be useful to have a label that matches a rating to a scraper/metadata provider?


> Wait if I'm not wrong is:
> Listitem.Rating(n)

i think you're settings properties? 
https://github.com/xbmc/xbmc/blob/master/xbmc/GUIInfoManager.cpp#L4987-L4998

phate89 commented:
No it's a specific property like art:
https://github.com/xbmc/xbmc/blob/master/xbmc/GUIInfoManager.cpp#L1447-L1459

I implemented it similarly to the art stuff. The basic idea is that a scraper like universal movie scraper can scrape more than one rating with a specific name. So for example if you enable imdb rating it will add a rating named "imdb" with the imdb rating value and votes. Of course we need an actual name to define what is from which site but it's the same for artwork. with "poster" you mean that specific artwork type...
Right now there's no point to link that rating to a scraper because the single movie info are only from one scraper.

MilhouseVH commented:
This is causing a crash due to recursion: http://sprunge.us/CQbD

If `m_strDefaultRating` is empty it will recurse forever and crash.

ronie commented:
> No it's a specific property like art

ah thanx, i get it now!
Listitem.Rating(imdb) to get the imdb rating, and so on..

anaconda commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -2340,6 +2434,9 @@ bool CVideoDatabase::UpdateDetailsForTvShow(int idTvShow, const CVideoInfoTag &d
   AddActorLinksToItem(idTvShow, MediaTypeTvShow, "director", details.m_director);
 
+  // add ratings
+  details.m_iIdRating = AddRatings(idTvShow, MediaTypeMovie, details.m_ratings, details.m_strDefaultRating);
```
`MediaTypeTvShow`

MilhouseVH commented:
With this PR, running `texturecache.py jd sets` will reliably crash kodi - this is executes the JSON query:

```
{"jsonrpc": "2.0", "params": {"sort": {"order": "ascending", "method": "title"}, "properties": ["title", "art"], "limits": {"start": 0, "end": 400}}, "method": "VideoLibrary.GetMovieSets", "id": "libSets"}
```

Full logs below from Ubuntu/x86_64.

kodi.log: http://sprunge.us/FHHQ
crashlog: http://sprunge.us/VEKJ

Without this PR, there are no problems.

phate89 commented:
@MilhouseVH  thanks for the test.
 They're both the same problem.
this should fix it: https://github.com/xbmc/xbmc/pull/8771

MilhouseVH commented:
Thanks, #8771 fixes it.

Tolriq commented:
@Montellese what is your decision about API versioning for Krypton ?

Now that this is merged and update to 7.1.0 seems we are doomed to handle anything and will surely see more major version update before release.

Razzeee commented:
This "broke" the testsuite.

The renaming of VIDEODB_ID_RATING, VIDEODB_ID_TV_RATING, VIDEODB_ID_EPISODE_RATING seem borked, I'LL come back with a fix

phate89 commented:
@Razzeee If I find the time and the will to do it, is a pr that remove the whole cXX stuff going to be accepted by the kodi team?

Razzeee commented:
@phate89 
absolutely, you'll get my full backup on that.
but it would be great to keep/update the database tests then.

phate89 commented:
sure.. I actually didn't notice the error because it compiled fine.. How can it compile fine with a missing variable? How can I trigger the tests from visual studio?

Razzeee commented:
Test don't run when you compile normally. 
In visual Studio you just need to change the config from debug or release to debug testsuite then run the solution. 
It's a drop down in the top menu bar.

anaconda commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -1385,6 +1410,67 @@ int CVideoDatabase::AddToTable(const std::string& table, const std::string& firs
+      {
+        id = m_pDS->fv(1).get_asInt();
+        m_pDS->close();
+        strSQL = PrepareSQL("UPDATE rating SET rating.rating = '%.f', rating.votes = %i WHERE rating.media_id = %i", i.second.rating, i.second.votes, id);
```
`id = m_pDS->fv(1).get_asInt();` should be `0` not `1`.

Query should be:
`strSQL = PrepareSQL("UPDATE rating SET rating = '%.f', votes = %i WHERE rating_id = %i", i.second.rating, i.second.votes, id);`

The combination of both typos produces the following errors:
```
00:14:18 2125.637695 T:1724208032   ERROR: SQL: [MyVideos102.db] SQL error or missing database
                                            Query: UPDATE rating SET rating.rating = '0', rating.votes = 0 WHERE rating.media_id = 0
00:14:18 2125.665527 T:1724208032   ERROR: AddRatings (674 - episode) failed
```

phate89 replied:
thanks for the catch.. I'll fix it immediately

anaconda commented on xbmc/interfaces/json-rpc/VideoLibrary.cpp:
```diff
@@ -1124,7 +1124,26 @@ void CVideoLibrary::UpdateVideoTag(const CVariant &parameterObject, CVideoInfoTa
     details.m_iTrack = (int)parameterObject["track"].asInteger();
   if (ParameterNotNull(parameterObject, "rating"))
-    details.m_fRating = parameterObject["rating"].asFloat();
+  {
```
Was there a reason for this change?
This change means one can no longer update just votes, but needs to update rating _and_ votes, so breaks existing tools for no apparent reason. I'd prefer to avoid this and just revert the grouping.
@phate89?

phate89 replied:
You are right again :) I'll fix it immediately

HitcherUK commented:
> ah thanx, i get it now!
> Listitem.Rating(imdb) to get the imdb rating, and so on..

Can't get this to work.

Tried `Listitem.Rating(imdb)` and `Listitem.Rating(default)` but both are empty.

phate89 commented:
@HitcherUK where did you try it? It's a detail like actors so it's available only in some case (like in info dialog)
